File: src/cli/commands/observability.ts
import { APP_ENV, envVarName } from "../../constants/env.js";
import {
  describeObservabilityModes,
  isThinkingEnabled,
  isVerboseEnabled,
  setThinkingOverride,
  setVerboseOverride,
} from "../../log/modes.js";
import { reply } from "./io.js";
import type { ReplCommandResult } from "./types.js";

function parseToggle(arg: string | undefined): boolean | null | "status" {
  if (!arg || arg === "status") {
    return "status";
  }
  if (arg === "on" || arg === "1" || arg === "true") {
    return true;
  }
  if (arg === "off" || arg === "0" || arg === "false") {
    return false;
  }
  return null;
}

export function handleThinkingCommand(arg: string | undefined): ReplCommandResult {
  const toggle = parseToggle(arg);
  if (toggle === null) {
    reply("usage: /thinking on|off|status");
    return "handled";
  }
  if (toggle === "status") {
    reply(describeObservabilityModes());
    reply(
      isThinkingEnabled()
        ? "shows trace call chain: thinking · tool → · result"
        : `thinking off — enable with /thinking on or ${envVarName(APP_ENV.THINKING)}=1`,
    );
    return "handled";
  }
  setThinkingOverride(toggle);
  reply(`thinking ${toggle ? "on" : "off"} · ${describeObservabilityModes()}`);
  return "handled";
}

export function handleVerboseCommand(arg: string | undefined): ReplCommandResult {
  const toggle = parseToggle(arg);
  if (toggle === null) {
    reply("usage: /verbose on|off|status");
    return "handled";
  }
  if (toggle === "status") {
    reply(describeObservabilityModes());
    reply(
      isVerboseEnabled()
        ? "verbose on — context · tool_use_log · conversation · full trace"
        : `verbose off — enable with /verbose on or ${envVarName(APP_ENV.VERBOSE)}=1`,
    );
    return "handled";
  }
  setVerboseOverride(toggle);
  reply(`verbose ${toggle ? "on" : "off"} · ${describeObservabilityModes()}`);
  return "handled";
}

File: crates/ocula-cli/src/commands/observability.rs
use std::sync::Arc;

use ocula_observability::{ObservabilityState, ToggleParse, parse_toggle};

pub fn handle_thinking_command(obs: &ObservabilityState, arg: Option<&str>) -> String {
    match parse_toggle(arg) {
        ToggleParse::Invalid => "usage: /thinking on|off|status".into(),
        ToggleParse::Status => {
            let mut lines = vec![obs.describe_modes()];
            lines.push(if obs.is_thinking_enabled() {
                "shows trace call chain: thinking · tool → · result".into()
            } else {
                "thinking off — enable with /thinking on or OCULA_THINKING=1".into()
            });
            lines.join("\n")
        }
        ToggleParse::On => {
            obs.set_thinking_override(Some(true));
            format!("thinking on · {}", obs.describe_modes())
        }
        ToggleParse::Off => {
            obs.set_thinking_override(Some(false));
            format!("thinking off · {}", obs.describe_modes())
        }
    }
}

pub fn handle_verbose_command(obs: &ObservabilityState, arg: Option<&str>) -> String {
    match parse_toggle(arg) {
        ToggleParse::Invalid => "usage: /verbose on|off|status".into(),
        ToggleParse::Status => {
            let mut lines = vec![obs.describe_modes()];
            lines.push(if obs.is_verbose_enabled() {
                "verbose on — context compose summary each turn".into()
            } else {
                "verbose off — enable with /verbose on or OCULA_VERBOSE=1".into()
            });
            lines.join("\n")
        }
        ToggleParse::On => {
            obs.set_verbose_override(Some(true));
            format!("verbose on · {}", obs.describe_modes())
        }
        ToggleParse::Off => {
            obs.set_verbose_override(Some(false));
            format!("verbose off · {}", obs.describe_modes())
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn thinking_status_does_not_panic() {
        let obs = Arc::new(ObservabilityState::default());
        let out = handle_thinking_command(&obs, Some("status"));
        assert!(out.contains("thinking:"));
    }

    #[test]
    fn verbose_on_implies_thinking_in_status() {
        let obs = Arc::new(ObservabilityState::default());
        let out = handle_verbose_command(&obs, Some("on"));
        assert!(out.contains("verbose on"));
        assert!(obs.is_thinking_enabled());
    }
}
